Sources/SwiftCompilerPlugin/Macros.swift

```diff
@@ -0,0 +1,254 @@
+      let diag = PluginMessage.Diagnostic(
+        from: .init(node: syntax, message: MacroExpansionError.freestandingMacroSyntaxIsNotMacro),
+        in: sourceManager
+      )
+      try pluginHostConnection.sendMessage(
+        .expandFreestandingMacroResult(
+          expandedSource: "",
+          diagnostics: [diag]
+        )
+      )
+      return
+    }
+
+    let macroDef = self.resolveMacro(macro)
+    ///
+    if let exprMacroDef = macroDef as? ExpressionMacro.Type {
+      let rewritten = try exprMacroDef.expansion(of: macroSyntax, in: context)
+      let diagnostics = context.diagnostics.map {
+        PluginMessage.Diagnostic(from: $0, in: sourceManager)
+      }
+
+      let resultMessage = PluginToHostMessage.expandFreestandingMacroResult(
+        expandedSource: rewritten.description,
+        diagnostics: diagnostics
+      )
+
+      try pluginHostConnection.sendMessage(resultMessage)
+      return
+    }
+
+    if let declMacroDef = macroDef as? DeclarationMacro.Type {
+      let rewritten = try declMacroDef.expansion(of: macroSyntax, in: context)
+      let diagnostics = context.diagnostics.map {
+        PluginMessage.Diagnostic(from: $0, in: sourceManager)
+      }
+
+      let resultMessage = PluginToHostMessage.expandFreestandingMacroResult(
+        expandedSource: rewritten.description,
+        diagnostics: diagnostics
+      )
+
+      try pluginHostConnection.sendMessage(resultMessage)
+      return
+    }
+
+    let diag = PluginMessage.Diagnostic(
+      from: Diagnostic(node: syntax, message: MacroExpansionError.macroTypeNotFound),
+      in: sourceManager
+    )
+
+    try pluginHostConnection.sendMessage(
+      .expandFreestandingMacroResult(expandedSource: "", diagnostics: [diag])
+    )
+  }
+
+  /// Expand `@attached(XXX)` macros.
+  func expandAttachedMacro(
+    macro: PluginMessage.MacroReference,
+    macroRole: PluginMessage.MacroRole,
+    discriminator: String,
+    customAttributeSyntax: PluginMessage.Syntax,
+    declSyntax: PluginMessage.Syntax,
+    parentDeclSyntax: PluginMessage.Syntax?
+  ) throws {
+    let sourceManager = SourceManager()
+    let context = PluginMacroExpansionContext(
+      sourceManager: sourceManager,
+      discriminator: discriminator
+    )
+
+    guard let macroDefinition = resolveMacro(macro) else {
+      fatalError("macro type not found: \(macro.moduleName).\(macro.typeName)")
+    }
+
+    let customAttributeNode = sourceManager.add(customAttributeSyntax).cast(AttributeSyntax.self)
+    let declarationNode = sourceManager.add(declSyntax).cast(DeclSyntax.self)
+
+    let expanded: [String]
+    switch (macroDefinition, macroRole) {
+    case (let attachedMacro as AccessorMacro.Type, .accessor):
+      let accessors = try attachedMacro.expansion(
+        of: customAttributeNode,
+        providingAccessorsOf: declarationNode,
+        in: context
+      )
+      expanded =
+        accessors
+        .map { $0.trimmedDescription }
```

Comment:
I think it would read better in one line but if you disagree, that’s also fine. Similar below.